Content.Server/Radio/EntitySystems/MessagesServerSystem.cs

```diff
@@ -0,0 +1,148 @@
+            toUpdate.Add((cartComponent.UserUid,cartComponent.UserName));
+        }
+
+        if (needDictUpdate)
+        {
+            query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+            while (query.MoveNext(out var cartUid,out var cartComponent))
+            {
+                foreach (var (key,value) in toUpdate)
+                {
+                    cartComponent.NameDict[key]=value;
+                }
+            }
+        }
+
+        if ((component.MessagesQueue.Count > 0))
+        {
+            var tempMessagesQueue = new List<MessagesMessageData>(component.MessagesQueue);
+            foreach (var message in tempMessagesQueue)
+            {
+                bool sent=TryToSend(message, mapId, component);
+                if (sent)
+                {
+                    component.MessagesQueue.Remove(message);
+                    component.Messages.Add(message);
+                }
+            }
+        }
+    }
+
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+        while(query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if ((cartComponent.UserUid == null) || (cartComponent.UserName == null))
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
+    public bool TryToSend(MessagesMessageData message, MapId mapId, MessagesServerComponent server)
+    {
+        bool sent = false;
+
+        var query = EntityQueryEnumerator<CartridgeLoaderComponent, ContainerManagerComponent>();
+
+        while (query.MoveNext(out var uid, out var comp, out var cont))
+        {
+            if (!_cartridgeLoaderSystem.TryGetProgram<MessagesCartridgeComponent>(uid, out var progUid, out var messagesCartridgeComponent, false, comp, cont))
+                continue;
+            if (progUid is EntityUid realProgUid)
+            {
+                if (messagesCartridgeComponent.UserUid == message.ReceiverId)
+                    _messagesCartridgeSystem.ServerToPdaMessage(realProgUid, messagesCartridgeComponent, message, uid, server);
+                sent = true;
+            }
+        }
+
+        return sent;
+    }
+
+    public void PdaToServerMessage(EntityUid uid, MessagesServerComponent component, MessagesMessageData message)
+    {
+        //component.Messages.Add(message);
+        component.MessagesQueue.Add(message);
+        Update(uid, component);
+    }
+
```

Comment:
Just iterate the component directly, there's no reason to be checking every single loader.